cpp/src/arrow/compute/row/grouper.cc

```diff
@@ -223,12 +223,11 @@ struct AnyKeysSegmenter : public BaseRowSegmenter {
 
   AnyKeysSegmenter(const std::vector<TypeHolder>& key_types, ExecContext* ctx)
       : BaseRowSegmenter(key_types),
-        ctx_(ctx),
-        grouper_(nullptr),
+        grouper_(Grouper::Make(key_types, ctx).ValueOrDie()),
```

Comment:
Is there a reasonable possibility that this might fail, for example due to input types?

Reply:
Without extra context, there is. But the factory method (cleverly) avoids this by a dummy instantiation of `Grouper`:
https://github.com/apache/arrow/blob/a715ea06b71ec206a987d7921264778e9954404b/cpp/src/arrow/compute/row/grouper.cc#L220

Reply:
But isn't it pointless to construct the `Grouper` twice? Can't the constructor simply take the `Grouper` that was constructed in the factory method?

Reply:
Ah, I see what you are saying. Will update. Thanks for pointing this out.

Reply:
Updated.